File: Java/src/main/java/Solution1001_1100/Solution1002.java
package Solution1001_1100;

import java.util.*;
import java.util.stream.Stream;

/**
 * @author windows
 */
public class Solution1002 {
    public List<String> commonChars(String[] words) {
        String str = words[0];
        for (int i = 1; i < words.length; i++) {
            HashMap<Character, Integer> map = new HashMap<>();
            for (int j = 0; j < str.length(); j++) {
                map.put(str.charAt(j), map.getOrDefault(str.charAt(j), 0) + 1);
            }
            StringBuilder sb = new StringBuilder();
            for (int j = 0; j < words[i].length(); j++) {
                char ch = words[i].charAt(j);
                if (map.containsKey(ch) && map.get(ch) > 0){
                    sb.append(ch);
                    map.put(ch, map.get(ch) - 1);
                }
            }
            str = sb.toString();
        }
        ArrayList<String> res = new ArrayList<>();
        for (int i = 0; i < str.length(); i++) {
            res.add(String.valueOf(str.charAt(i)));
        }
        return res;
    }
    
    
    public static void main(String[] args) {
        Solution1002 solution1002 = new Solution1002();
    }
}

File: C++/src/Solution1001_1100/Solution1002.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution1002.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;


vector<string> commonChars(vector<string>& words) {
   vector<int> chars(26);
   for (const auto &item : words[0]){
     chars[item - 'a']++;
   }
   for (int i = 1; i < words.size(); ++i) {
     vector<int> charsTemp(26);
     for (const auto &item : words[i]){
       charsTemp[item - 'a']++;
     }
     for (int j = 0; j < 26; ++j) {
       chars[j] = min(chars[j], charsTemp[j]);
     }
   }

   vector<string> res;
   for (int i = 0; i < 26; ++i) {
     for (int j = 0; j < chars[i]; ++j) {
       res.push_back(string(1, 'a' + i));
     }
   }
   return res;
}
